streams/src/main/java/org/apache/kafka/streams/processor/internals/metrics/ThreadMetrics.java

```diff
@@ -0,0 +1,179 @@
+                                  COMMIT_TOTAL_DESCRIPTION,
+                                  COMMIT_RATE_DESCRIPTION);
+        return commitSensor;
+    }
+
+    public static Sensor pollSensor(final StreamsMetricsImpl streamsMetrics) {
+        final Sensor pollSensor = streamsMetrics.threadLevelSensor(POLL, Sensor.RecordingLevel.INFO);
+        final Map<String, String> tagMap = streamsMetrics.threadLevelTagMap();
+        addAvgAndMax(pollSensor, THREAD_LEVEL_GROUP, tagMap, POLL_LATENCY);
+        addInvocationRateAndCount(pollSensor,
+                                  THREAD_LEVEL_GROUP,
+                                  tagMap,
+                                  POLL,
+                                  POLL_TOTAL_DESCRIPTION,
+                                  POLL_RATE_DESCRIPTION);
+        return pollSensor;
+    }
+
+    public static Sensor processSensor(final StreamsMetricsImpl streamsMetrics) {
+        final Sensor processSensor = streamsMetrics.threadLevelSensor(PROCESS, Sensor.RecordingLevel.INFO);
+        final Map<String, String> tagMap = streamsMetrics.threadLevelTagMap();
+        addAvgAndMax(processSensor, THREAD_LEVEL_GROUP, tagMap, PROCESS_LATENCY);
+        addInvocationRateAndCount(processSensor,
+                                  THREAD_LEVEL_GROUP,
+                                  tagMap,
+                                  PROCESS,
+                                  PROCESS_TOTAL_DESCRIPTION,
+                                  PROCESS_RATE_DESCRIPTION);
+
+        return processSensor;
+    }
+
+    public static Sensor punctuateSensor(final StreamsMetricsImpl streamsMetrics) {
+        final Sensor punctuateSensor = streamsMetrics.threadLevelSensor(PUNCTUATE, Sensor.RecordingLevel.INFO);
+        final Map<String, String> tagMap = streamsMetrics.threadLevelTagMap();
+        addAvgAndMax(punctuateSensor, THREAD_LEVEL_GROUP, tagMap, PUNCTUATE_LATENCY);
+        addInvocationRateAndCount(punctuateSensor,
+                                  THREAD_LEVEL_GROUP,
+                                  tagMap,
+                                  PUNCTUATE,
+                                  PUNCTUATE_TOTAL_DESCRIPTION,
+                                  PUNCTUATE_RATE_DESCRIPTION);
+
+        return punctuateSensor;
+    }
+
+    public static Sensor skipRecordSensor(final StreamsMetricsImpl streamsMetrics) {
+        final Sensor skippedRecordsSensor = streamsMetrics.threadLevelSensor(SKIP_RECORD, Sensor.RecordingLevel.INFO);
+        addInvocationRateAndCount(skippedRecordsSensor,
+                                  THREAD_LEVEL_GROUP,
+                                  streamsMetrics.threadLevelTagMap(),
+                                  SKIP_RECORD,
+                                  SKIP_RECORD_TOTAL_DESCRIPTION,
+                                  SKIP_RECORD_RATE_DESCRIPTION);
+
+        return skippedRecordsSensor;
+    }
+
+    public static Sensor commitOverTasksSensor(final StreamsMetricsImpl streamsMetrics) {
```

Comment:
I'm a bit confused with this `commitOverTasksSensor` and `commitSensor`: they are both on thread-level, and the `commitOverTasksSensor` seems only be used as the parent of task-level `commitSensor`.

The parent of it was defined as task-level sensor still, while the only difference is that the `taskId` tag is `all`. Is there any changes intentional here?

Reply:
@guozhangwang, I checked `commitOverTasksSensor` and `commitSensor`. First of all, I have not changed anything in this PR regarding the public interface, i.e. tags and groups. 

From an implementation point of view, in this PR `commitOverTasksSensor` is created and added to the metrics on thread-level, which I think is better as the current implementation that relies on the idempotency of adding a sensor and adds each time a task is created (cf. https://github.com/apache/kafka/blob/b042b36674f0bd8046f0229c96e0b8cf99554b2b/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L99-L116).

However, now I am also a bit confused about `commitSensor` and `commitOverTasksSensor`. `commitSensor` measures commits on thread-level whereas `commitOverTasksSensor` aggregates measurements of commits on task-level. Hence, `commitSensor` approximately subsumes `commitOverTasksSensor`. Furthermore, `commitSensor` does also measure commits of stand-by tasks. Looking at KIP-444, we will get rid of `commitOverTasksSensor`.

Reply:
Yeah I know that KIP-444 would only maintain the `commitSensor` on thread-level, and on task-level, and these two would not have a parent-children relationship any more. What I'm just confused is if we have ever changed any protocol in this PR. As long as nothing is changed I think it is okay -- and probably we just had some old error code which we are going to get rid of in KIP-444 anyways.

Let me do a double check and if that confirms my conjecture, we can merge it as is.